docs/core/compatibility/aspnet-core/7.0/signalr-hub-method-parameters-di.md

```diff
@@ -0,0 +1,82 @@
+---
+title: "Breaking change: SignalR Hub methods try to resolve parameters from DI"
+description: Learn about the breaking change in ASP.NET Core 7.0 where SignalR Hub methods try to resolve parameters from DI.
+ms.date: 03/03/2022
+---
+
+# SignalR Hub methods try to resolve parameters from DI
+
+SignalR Hub methods now support injecting services from your Dependency Injection (DI) container. In rare cases, this can break apps that have a type in DI that is also accepted in Hub methods from SignalR client messages.
+
+## Version introduced
+
+ASP.NET Core 7.0 Preview 2
+
+## Previous behavior
+
+If you accepted a type in your Hub method that was also in your Dependency Injection container, the type would always be resolved from a message sent by the client.
+
+```csharp
+Services.AddScoped<SomeCustomType>();
+
+class MyHub : Hub
+{
+    // type always comes from the client, never comes from DI
+    public Task Method(string text, SomeCustomType type) => Task.CompletedTask;
+}
+```
+
+## New behavior
+
+Types in DI are checked at app startup using <xref:Microsoft.Extensions.DependencyInjection.IServiceProviderIsService> to determine if an argument in a Hub method comes from DI or from the client.
+
+In the following example, which assumes you're using the default DI container, `SomeCustomType` comes from the DI container instead of from the client. If the client tries to send `SomeCustomType`, it gets an error:
+
+```csharp
+Services.AddScoped<SomeCustomType>();
+
+class MyHub : Hub
+{
+    // comes from DI by default
+    public Task Method(string text, SomeCustomType type) => Task.CompletedTask;
+}
+```
+
+## Type of breaking change
+
+This change affects [source compatibility](../../categories.md#source-compatibility).
+
+## Reason for change
+
+We believe the likelihood of breaking apps to be very low as it isn't common to have a type in DI and as an argument in your Hub method at the same time.
```

Comment:
This isn't really a "reason for change"; nevertheless, I left a small suggestion.

```suggestion
The likelihood of breaking apps is low as it isn't common to have a type in DI and as an argument in your Hub method at the same time.
```

Reply:
Isn't the "reason for change" in the context of making a breaking change? So this seems perfectly reasonable as the reason we were fine making a breaking change.

Reply:
That's not how I've seen this field used in the .NET breaking changes. I think of it more like "why it improves the experience". For example, [this issue](https://github.com/dotnet/docs/issues/28372).

Reply:
What should I do with this one? A similar announcement for DI in MVC's API controllers has this:

> This same behavior is already implemented in minimal APIs.
>
> The likelihood of breaking apps is low as it isn't common to have a type in DI and as an argument in your API controller action at the same time.

I see that as covering both why the change was made and why the change was acceptable as being breaking. Would you like to add a line similar to the first one in the example I gave above?